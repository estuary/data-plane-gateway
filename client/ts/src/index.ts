export * as protocols from "./protocols.ts";
export * as streams from "./streams.ts";
export { JournalClient } from "./journal_client.ts";
export { JournalSelector, ShardSelector } from "./selector.ts";
export { Result } from "./result.ts";
export { ShardClient } from "./shard_client.ts";
