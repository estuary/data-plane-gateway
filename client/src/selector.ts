import * as broker from "./gen/broker/protocol/broker.ts";
import { sortBy } from "./util.ts";

export class Selector {
  protected labels: Array<broker.ProtocolLabel>;

  constructor() {
    this.labels = [];
  }

  toLabelSet(): broker.ProtocolLabelSet {
    const sorted = this.labels.slice(0).sort(sortBy("name", "value"));
    return { labels: sorted };
  }

}

export class JournalSelector extends Selector {
  collection(v: string): this {
    this.labels.push({ name: "estuary.dev/collection", value: v });
    return this;
  }

  name(v: string): this {
    this.labels.push({ name: "name", value: v });
    return this;
  }

  static prefix(v: string): JournalSelector {
    if (!v.endsWith("/")) {
      v = v + "/";
    }
    const sel = new JournalSelector();
    sel.labels.push({ name: "prefix", value: v });
    return sel;
  }
}

export class ShardSelector extends Selector {
  task(v: string): this {
    this.labels.push({ name: "estuary.dev/task-name", value: v });
    return this;
  }

  id(v: string): this {
    this.labels.push({ name: "id", value: v });
    return this;
  }
}
