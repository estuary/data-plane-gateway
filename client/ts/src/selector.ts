import { broker } from "./protocols.ts";

export class Selector {
  protected labels: Array<broker.ProtocolLabel>;

  constructor() {
    this.labels = [];
  }

  toLabelSet(): broker.ProtocolLabelSet {
    return { labels: this.labels };
  }
}

export class JournalSelector extends Selector {
  collection(v: string): this {
    this.labels.push({ name: "estuary.dev/collection", value: v });
    return this;
  }

  name(v: string): this {
    this.labels.push({ name: "name", value: v });
    return this;
  }

  static prefix(v: string): JournalSelector {
    if (!v.endsWith("/")) {
      v = v + "/";
    }
    const sel = new JournalSelector();
    sel.labels.push({ name: "prefix", value: v });
    return sel;
  }
}

export class ShardSelector extends Selector {
  task(v: string): this {
    this.labels.push({ name: "estuary.dev/task-name", value: v });
    return this;
  }

  id(v: string): this {
    this.labels.push({ name: "id", value: v });
    return this;
  }
}
