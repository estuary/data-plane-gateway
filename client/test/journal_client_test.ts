import {
  assertEquals,
  assertMatch,
} from "https://deno.land/std@0.135.0/testing/asserts.ts";
// TODO: Built-in snapshot testing has landed in deno std, but is not yet released.
import { test as snapshotTest } from "https://deno.land/x/snap/mod.ts";

import * as broker from "../src/gen/broker/protocol/broker.ts";
import { BASE_URL, makeJwt } from "./test_support.ts";
import { JournalClient, parseJournalDocuments } from "../src/journal_client.ts";
import { JournalSelector } from "../src/selector.ts";
import { readStreamToEnd } from "../src/streams.ts";

Deno.test("JournalClient.list collection selector test", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const name = new JournalSelector().collection("acmeCo/greetings");
  const journals = (await client.list(name)).unwrap();

  assertEquals(1, journals.length);
  assertEquals("acmeCo/greetings/pivot=00", journals[0].name);
});

Deno.test("JournalClient.list name selector test", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const name = new JournalSelector().name("acmeCo/greetings/pivot=00");
  const journals = (await client.list(name)).unwrap();

  assertEquals(1, journals.length);
  assertEquals("acmeCo/greetings/pivot=00", journals[0].name);
});

Deno.test("JournalClient.list prefix selector test", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));
  const expectedJournals = [
    "ops/acmeCo/logs/kind=capture/name=acmeCo%2Fsource-hello-world/pivot=00",
    "ops/acmeCo/stats/kind=capture/name=acmeCo%2Fsource-hello-world/pivot=00",
  ];

  const prefixSelector = JournalSelector.prefix("ops/acmeCo/");
  const journals = (await client.list(prefixSelector)).unwrap();

  assertEquals(2, journals.length);
  assertEquals(expectedJournals, journals.map((j) => j.name).sort());
});

Deno.test("JournalClient.list exclusion selector test", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const prefixSelector = JournalSelector.prefix("ops/acmeCo/");
  const excludedSelector = new JournalSelector().collection("ops/acmeCo/stats");
  const journals = (await client.list(prefixSelector, excludedSelector))
    .unwrap();

  assertEquals(1, journals.length);
  assertEquals(
    "ops/acmeCo/logs/kind=capture/name=acmeCo%2Fsource-hello-world/pivot=00",
    journals[0].name,
  );
});

Deno.test("JournalClient.list wrong signing key", async () => {
  const mismatchedKey = await makeJwt({
    key: new TextEncoder().encode("wrong key"),
  });
  const client = new JournalClient(BASE_URL, mismatchedKey);

  const name = new JournalSelector().collection("acmeCo/greetings");
  const error = (await client.list(name)).unwrap_err();

  assertMatch(error.body.message!, new RegExp("signature is invalid"));
});

Deno.test("JournalClient.list auth expiration", async () => {
  const expired_jwt = await makeJwt({ expiresAt: 1640995200 }); // Midnight Jan 1, 2022
  const client = new JournalClient(BASE_URL, expired_jwt);

  const name = new JournalSelector().collection("acmeCo/greetings");
  const error = (await client.list(name)).unwrap_err();

  assertMatch(error.body.message!, new RegExp("token is expired"));
});

Deno.test("JournalClient.list unauthorized prefix", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const name = new JournalSelector()
    // An authorized selector...
    .collection("acmeCo/greetings")
    // ...as well as an unauthorized selector
    .collection("wayneEnterprises/batcave/sensors");
  const error = (await client.list(name)).unwrap_err();

  assertMatch(
    error.body.message!,
    new RegExp("unauthorized `estuary.dev/collection` label"),
  );
});

snapshotTest("JournalClient.read test", async ({ assertSnapshot }) => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const req = { journal: "acmeCo/greetings/pivot=00", endOffset: "1024" };
  const stream = (await client.read(req)).map_err(console.error).unwrap();
  const results: Array<broker.ProtocolReadResponse> = await readStreamToEnd(
    stream,
  );

  // Pluck out a few interesting properties to snapshot on. Many of the
  // properties are not stable between clusters, so snapshotting the entire
  // response would involve redacting a lot of fields.
  const pluck = (res: broker.ProtocolReadResponse) => {
    return {
      status: res.status,
      fragment: {
        journal: res!.fragment?.journal,
        compressionCodec: res!.fragment?.compressionCodec,
      },
    };
  };

  assertSnapshot(results.map(pluck));
});

snapshotTest("JournalClient.read content test", async ({ assertSnapshot }) => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const req = {
    journal: "acmeCo/greetings/pivot=00",
    offset: "10",
    endOffset: "1024",
  };
  const stream = (await client.read(req)).unwrap();
  const docStream = parseJournalDocuments(stream!);
  const results = await readStreamToEnd(docStream);

  let filtered_results = results.filter(r=>!(r._meta as any).ack)

  const masks = ["/*/_meta/uuid"];
  assertSnapshot(filtered_results, masks);
});

Deno.test("JournalClient.read unauthorized prefix", async () => {
  const client = new JournalClient(BASE_URL, await makeJwt({}));

  const req = {
    journal: "wayneEnterprises/batcave/sensors",
    endOffset: "1024",
  };
  const error = (await client.read(req)).unwrap_err();

  assertMatch(error.body.message!, new RegExp("was not found in claims"));
});
