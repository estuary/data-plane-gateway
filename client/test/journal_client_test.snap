{
  'JournalClient.read content test': [
    {
      _meta: {
        uuid: '[MASKED string]',
      },
      message: 'Hello #1',
      ts: '[MASKED string]',
    },
    {
      _meta: {
        uuid: '[MASKED string]',
      },
      message: 'Hello #2',
      ts: '[MASKED string]',
    },
    {
      _meta: {
        uuid: '[MASKED string]',
      },
      message: 'Hello #3',
      ts: '[MASKED string]',
    },
    {
      _meta: {
        uuid: '[MASKED string]',
      },
      message: 'Hello #4',
      ts: '[MASKED string]',
    },
    {
      _meta: {
        uuid: '[MASKED string]',
      },
      message: 'Hello #5',
      ts: '[MASKED string]',
    },
  ],
  'JournalClient.read test': [
    {
      status: 'OK',
      fragment: {
        journal: 'acmeCo/greetings/pivot=00',
        compressionCodec: 'GZIP',
      },
    },
    {
      status: 'OK',
      fragment: {},
    },
    {
      status: 'OK',
      fragment: {
        journal: 'acmeCo/greetings/pivot=00',
        compressionCodec: 'GZIP',
      },
    },
    {
      status: 'OK',
      fragment: {},
    },
  ],
}